arch/risc-v/src/bl808/bl808_wdt.c

```diff
@@ -0,0 +1,448 @@
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/arch.h>
+
+#include <sys/types.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <debug.h>
+
+#include <nuttx/irq.h>
+#include <nuttx/timers/watchdog.h>
+#include <nuttx/fs/ioctl.h>
+
+#include "hardware/bl808_timer.h"
+#include "riscv_internal.h"
+#include "chip.h"
+#include "bl808_wdt.h"
+
+#ifdef CONFIG_BL808_WDT
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define WDT_CLK_SRC_1K 2
+#define WDT_CLK_SRC_NONE 5
+
+#define BL808_UNLOCK_WDT(n) \
+  ({ \
+    putreg32(0xbaba, BL808_WDT_KEY1(n)); \
+    putreg32(0xeb10, BL808_WDT_KEY2(n)); \
+  })
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct bl808_wdt_s
+{
+  const struct watchdog_ops_s *ops;
+  int idx;
+  xcpt_t callback;
+  bool started;
+  uint32_t timeout;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+int bl808_wdt_start(FAR struct watchdog_lowerhalf_s *lower);
+int bl808_wdt_stop(FAR struct watchdog_lowerhalf_s *lower);
+int bl808_wdt_keepalive(FAR struct watchdog_lowerhalf_s *lower);
+int bl808_wdt_getstatus(FAR struct watchdog_lowerhalf_s *lower,
+                          FAR struct watchdog_status_s *status);
+int bl808_wdt_settimeout(FAR struct watchdog_lowerhalf_s *lower,
+                           uint32_t timeout);
+xcpt_t bl808_wdt_capture(FAR struct watchdog_lowerhalf_s *lower,
+                       CODE xcpt_t callback);
+int bl808_wdt_ioctl(FAR struct watchdog_lowerhalf_s *lower,
+                      int cmd, unsigned long arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static struct watchdog_ops_s bl808_wdt_ops =
+  {
+    .start = bl808_wdt_start,
+    .stop = bl808_wdt_stop,
+    .keepalive = bl808_wdt_keepalive,
+    .getstatus = bl808_wdt_getstatus,
+    .settimeout = bl808_wdt_settimeout,
+    .capture = bl808_wdt_capture,
+    .ioctl = bl808_wdt_ioctl,
+  };
+
+static struct bl808_wdt_s wdt0 =
+  {
+    .ops = &bl808_wdt_ops,
+    .idx = 0,
+    .callback = NULL,
+    .started = false,
+    .timeout = 0
+  };
+
+static struct bl808_wdt_s wdt1 =
+  {
+    .ops = &bl808_wdt_ops,
+    .idx = 1,
+    .callback = NULL,
+    .started = false,
+    .timeout = 0
+  };
```

Comment:
I'm not sure if it is stated in code style guidelines, but I think that usually we do not have 2 spaces indentation for struct init blocks in other places.